src/shogun/metric/LMNN.cpp

```diff
@@ -104,13 +104,10 @@ void CLMNN::train(SGMatrix<float64_t> init_transform)
 
 		// Compute the objective, trace of Mahalanobis distance matrix (L squared) times the gradient
 		// plus the number of current impostors to account for the margin
-		// trace of the product of two matrices computed fast using
-		// trace(A*B)=sum(A.*B')
 		SG_DEBUG("Computing objective.\n")
 		obj[iter] = m_regularization * cur_impostors.size();
-		obj[iter] += linalg::trace(
-		    linalg::element_prod(
-		        linalg::matrix_prod(L, L, true, false), gradient));
+		obj[iter] +=
```

Comment:
Is this the same as before? Doesnt seem to me but maybe I am blind :)

Reply:
They are different. Actually I was wrong in #4167. Now it is the same as the original one.  https://github.com/shogun-toolbox/shogun/blob/431e56b67344e0662444c0efbd4aefde47990fa4/src/shogun/metric/LMNN.cpp#L113

Reply:
EHM!!!
Why did the tests pass? Or are there none?

Reply:
Well, here the trace is used as the termination condition. Seems that it doesn't stop until reaching maximum number of iterations when trace is wrong in tests.

Reply:
I see now! makes sense
Ok, this definitely should be tested somehow (at least in a unit test)
Wanna do send another PR for that?

Reply:
integration tests are back, so now let's see :)

Reply:
Hey guys. Why isn't the same as before? I don't see it :-/ Or maybe this discussion is about a previous diff?

Reply:
Check the history of that line, it changed back and forth IIRC

Reply:
I checked and only found the three mentioned here: the original with TRACE plust the two in this diff.

Reply:
The original with TRACE is correct, the former one in this diff is wrong.

Reply:
Where is the difference between them, @vinx13?

Reply:
We are calculating trace of L' * L * gradient, which is sum( (L'*L) .* gradient' ), instread of trace( (L'*L) .* gradient) that appears in the previous commit